sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java

```diff
@@ -300,7 +300,8 @@ private void unpackMiniBlock() throws IOException {
         bitWidths[currentMiniBlock]);
     for (int j = 0; j < miniBlockSizeInValues; j += 8) {
       ByteBuffer buffer = in.slice(packer.getBitWidth());
-      packer.unpack8Values(buffer, buffer.position(), unpackedValuesBuffer, j);
+      packer.unpack8Values(buffer.array(),
```

Comment:
I think it's better to check `buffer.hasArray` before calling this, since `parquet-mr` may not necessarily return heap buffer all the time.

Reply:
@sunchao Double check the code path,  `in` is an `ByteBufferInputStream` and passed in by `initFromPage`  method.

The `ByteBufferInputStream` here should come from

https://github.com/apache/spark/blob/bf8a4c47ac3752edf86f1e14e4050e8d202b34a4/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java#L389

`page.getData().toInputStream()` always return a `SingleBufferInputStream` wrapped `HeapByteBuffer`, `SingleBufferInputStream. slice` should still return a `HeapByteBuffer`, so  in the current code path, `buffer.hasArray` should always return true. If I'm wrong, please correct me.

Reply:
Yes, that's the current state, but it'll subject to change though. I know there are some discussion to make more use of direct byte buffer in `parquet-mr`. Therefore, it'd be better to make this code more future proof.

Reply:
(In fact, we've already changed this internally to return direct byte buffer if some flag is turned on).

Reply:
Got it, let me think about how to reduce the number of calls to `buffer.hasArray`

Reply:
Since it's different from AS-IS code, it would be great if we can add a comment about the official Parquet JIRA issue. Could you give me the pointer, @sunchao ?

Reply:
BTW, I believe it's worth to have another independent JIRA, @sunchao and @LuciferYang . Please file a new one and link to Parquet JIRA which @sunchao is going to provide.

Reply:
> @LuciferYang , please file a new one and link to Parquet JIRA which @sunchao is going to provide.

OK

Reply:
there is no parquet JIRA yet

Reply:
Could you file one for trace-ability in order not to forget, @sunchao ?